components/board/genericlinux/gpio.go

```diff
@@ -210,27 +255,39 @@ func (pin *gpioPin) Close() error {
 	pin.mu.Lock()
 	defer pin.mu.Unlock()
 
+	if pin.hwPwm != nil {
+		// Make sure to unexport the sysfs device for hardware PWM on this pin, if it's in use.
+		if err := pin.hwPwm.Close(); err != nil {
+			return err
+		}
+	}
 	if pin.line == nil {
 		return nil // Never opened, so no need to close
 	}
 
-	err := pin.line.Close()
-	pin.line = nil
-	return err
+	return pin.closeGpioFd()
 }
 
 func gpioInitialize(cancelCtx context.Context, gpioMappings map[int]GPIOBoardMapping,
 	waitGroup *sync.WaitGroup, logger golog.Logger,
 ) map[string]*gpioPin {
 	pins := make(map[string]*gpioPin)
-	for pin, mapping := range gpioMappings {
-		pins[fmt.Sprintf("%d", pin)] = &gpioPin{
+	for pinNumber, mapping := range gpioMappings {
+		pin := &gpioPin{
 			devicePath: mapping.GPIOChipDev,
 			offset:     uint32(mapping.GPIO),
 			cancelCtx:  cancelCtx,
 			waitGroup:  waitGroup,
 			logger:     logger,
 		}
+		if mapping.HWPWMSupported {
+			hwPwm, err := NewPwmDevice(mapping.PWMSysFsDir, mapping.PWMID)
+			if err != nil {
+				return pins // TODO: replace with (nil, err) after merging other PR.
```

Comment:
An explanation of this comment: we ought to change this function signature to also return an error, but we're already making that change over in https://github.com/viamrobotics/rdk/pull/2000 and doing it again here will lead to nasty merge conflicts. So, we'll wait until that PR is merged, then rebase this one and fix this TODO.